pkg/s3wrapper/util.go

```diff
@@ -23,3 +34,180 @@ func FixEndpointURL(endpoint string) (string, error) {
+		exists, err := api.DoesObjectExist(ctx, objectName)
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to check existence of %s", objectName)
+			log.Error(err)
+			return err
+		}
+		if exists {
+			continue
+		}
+		if !haveISOFile {
+			// We need to upload a boot file but don't have it locally
+			tmpFile, fileErr := ioutil.TempFile(os.TempDir(), "tmpiso-")
+			if fileErr != nil {
+				fileErr = errors.Wrapf(fileErr, "Failed to create temp file in directory %s", os.TempDir())
+				log.Error(fileErr)
+				return fileErr
+			}
+			defer os.Remove(tmpFile.Name())
+
+			fileErr = downloadToFile(RHCOSBaseISOURL, tmpFile)
+			if fileErr != nil {
+				log.Error(fileErr)
+				return fileErr
+			}
+
+			isoFilePath = tmpFile.Name()
+			haveISOFile = true
+
+		}
+		if fileType == "iso" {
+			err = uploadLeader.RunWithLeader(ctx, func() error {
+				log.Infof("Starting upload of ISO %s", objectName)
+				return api.UploadFile(ctx, isoFilePath, objectName)
+			})
+		} else {
+			err = uploadLeader.RunWithLeader(ctx, func() error {
+				log.Infof("Starting upload of file %s", objectName)
+				return uploadFileFromISO(ctx, isoFilePath, fileType, objectName, api)
+			})
+		}
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to upload boot file %s", objectName)
+			log.Error(err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func downloadToFile(url string, destFile *os.File) error {
+	resp, err := http.Get(url)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to access URL %s", url)
+	}
+	defer resp.Body.Close()
+	defer destFile.Close()
+
+	if _, err = io.Copy(destFile, resp.Body); err != nil {
+		return errors.Wrapf(err, "Failed to download URL to %s", destFile.Name())
+	}
+	return nil
+}
+
+func getVolumeIDFromISO(isoFilePath string) (string, error) {
+	iso, err := os.Open(isoFilePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to open provided base image for inspection: %s", isoFilePath)
+	}
+	defer iso.Close()
+
+	// Need a method to identify the ISO provided
+	// Based on the ISO 9660 standard we should consistently be able to
+	// grab the Volume ID here.
+	volumeId := make([]byte, 32)
+	_, err = iso.ReadAt(volumeId, 32808)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to read volume ID from provided base image")
+	}
+
+	// TODO: Should we verify that the volume id is of the form `rhcos-<version>`?
+	return strings.TrimSpace(string(volumeId)), nil
+}
+
+func getBootFileName(isoObjectName, fileType string) string {
+	return strings.TrimSuffix(isoObjectName, ".iso") + "." + fileType
+}
+
+func uploadFileFromISO(ctx context.Context, isoFilePath, fileType, objectName string, api API) error {
+	filename, err := convertBootFileTypeToFileName(fileType)
+	if err != nil {
+		return err
+	}
+	f, err := os.Open(isoFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to open ISO file: %s", isoFilePath)
+	}
+	defer f.Close()
+	image, err := iso9660.OpenImage(f)
```

Comment:
Want to use https://github.com/openshift/assisted-service/pull/803 for this? :pray: :sweat_smile: